File: leetcode/array/1122. Relative Sort Array/py/main.py
import collections

# Countion sort
# Time O(1)
# Space O(1)
class Solution:
    def relativeSortArray(self, arr1, arr2):
        count = [0] * 1001 
        
        for x in arr1:
            count[x] += 1
        
        index = 0

        for num in arr2:
          while count[num] > 0:
            arr1[index] = num
            index += 1
            count[num] -= 1
        
        for num in range(1001):
            while count[num] > 0:
                arr1[index] = num
                index += 1
                count[num] -= 1
        
        return arr1

File: leetcode/array/1122. Relative Sort Array/js/main.js
// Time O(N)
// Space O(N)
const relativeSortArray = (arr1, arr2) => {
  let map = {};
  for (let x of arr1) {
    map[x] = ~~map[x] + 1;
  }

  let pos = 0;
  for (let x of arr2) {
    while (map[x]-- > 0) arr1[pos++] = x;
  }

  for (let x of Object.keys(map)) {
    while (map[x]-- > 0) arr1[pos++] = x;
  }

  return arr1;
};

// Time O(N)
// Space O(N)
const relativeSortArray = (arr1, arr2) => {
  let map = new Map();

  for (let i = 0; i < arr2.length; i++) {
    map.set(arr2[i], i);
  }

  let first = arr1.filter(x => map.has(x)).sort((a, b) => map.get(a) - map.get(b));
  let second = arr1.filter(x => !map.has(x)).sort((a, b) => a - b);

  return [...first, ...second];
};
